src/theory/quantifiers/sygus/sygus_eval_unfold.cpp

```diff
@@ -0,0 +1,206 @@
+
+using namespace std;
+using namespace CVC4::kind;
+using namespace CVC4::context;
+
+namespace CVC4 {
+namespace theory {
+namespace quantifiers {
+
+SygusEvalUnfold::SygusEvalUnfold(TermDbSygus* tds) : d_tds(tds) {}
+
+void SygusEvalUnfold::registerEvalTerm(Node n)
+{
+  Assert(options::sygusEvalUnfold());
+  // is this an APPLY_UF term with head that is a sygus datatype term?
+  if (n.getKind() != APPLY_UF)
+  {
+    return;
+  }
+  TypeNode tn = n[0].getType();
+  if (!tn.isDatatype())
+  {
+    return;
+  }
+  const Datatype& dt = static_cast<DatatypeType>(tn.toType()).getDatatype();
+  if (!dt.isSygus())
+  {
+    return;
+  }
+  Node f = n.getOperator();
+  if (n[0].getKind() == APPLY_CONSTRUCTOR)
+  {
+    // constructors should be unfolded and reduced already
+    return;
+  }
+  if (d_eval_processed.find(n) != d_eval_processed.end())
+  {
+    return;
+  }
+  Trace("sygus-eval-unfold")
+      << "SygusEvalUnfold: register eval term : " << n << std::endl;
+  d_eval_processed.insert(n);
+  // is it the sygus evaluation function?
+  Node eval_op = Node::fromExpr(dt.getSygusEvaluationFunc());
+  if (n.getOperator() != eval_op)
+  {
+    return;
+  }
+  // register this evaluation term with its head
+  d_evals[n[0]].push_back(n);
+  Node var_list = Node::fromExpr(dt.getSygusVarList());
+  d_eval_args[n[0]].push_back(std::vector<Node>());
+  for (unsigned j = 1, size = n.getNumChildren(); j < size; j++)
+  {
+    d_eval_args[n[0]].back().push_back(n[j]);
+  }
+  Node a = TermDbSygus::getAnchor(n[0]);
+  d_subterms[a].insert(n[0]);
+}
+
+void SygusEvalUnfold::registerModelValue(Node a,
+                                         Node v,
+                                         std::vector<Node>& terms,
+                                         std::vector<Node>& vals,
+                                         std::vector<Node>& exps)
+{
+  std::map<Node, std::unordered_set<Node, NodeHashFunction> >::iterator its =
+      d_subterms.find(a);
+  if (its == d_subterms.end())
+  {
+    return;
+  }
+  SygusExplain* sy_exp = d_tds->getExplain();
+  Trace("sygus-eval-unfold")
+      << "SygusEvalUnfold: " << a << ", has " << its->second.size()
+      << " registered subterms." << std::endl;
+  for (const Node& n : its->second)
+  {
+    Trace("sygus-eval-unfold-debug") << "...process : " << n << std::endl;
+    std::map<Node, std::vector<std::vector<Node> > >::iterator it =
+        d_eval_args.find(n);
+    if (it != d_eval_args.end() && !it->second.empty())
+    {
+      TNode at = a;
+      TNode vt = v;
+      Node vn = n.substitute(at, vt);
+      vn = Rewriter::rewrite(vn);
+      unsigned start = d_node_mv_args_proc[n][vn];
+      // get explanation in terms of testers
+      std::vector<Node> antec_exp;
+      sy_exp->getExplanationForEquality(n, vn, antec_exp);
+      Node antec = antec_exp.size() == 1
+                       ? antec_exp[0]
+                       : NodeManager::currentNM()->mkNode(kind::AND, antec_exp);
```

Comment:
cache NodeManager and also do not use kind prefix since you are in kind namespace